rfcs/unboxed-types.md

```diff
@@ -0,0 +1,2036 @@
+    update, even if it is potentially non-atomic. The inferred layout of a most
+    (see below) product types (like `#t`) will be built with `*w`, not `*`. Yet
+    if a type is exported abstractly from a module, the module signature will
+    have to specify the layout of that type; most users will write a layout
+    using `*`, thus protecting values of that type from getting torn.
+
+    If a user wants to write a layout with `*w`, they can do so like this:
+
+    ```ocaml
+    type t : bits32 * (bits32 * bits32) [@writable]
+    ```
+
+    The `[@writable]` attribute transforms all `*`s in a layout to be `*w`s.
+
+    Despite most product types being inferred to have a `*w` layout, product
+    types with existential variables are inferred to have a `*` layout, as
+    tearing an existential could be type-unsafe. Here is an example:
+
+    ```ocaml
+    type ex_int = K : 'a * ('a -> int) -> ex_int
+    ```
+
+    The layout of `#ex_int` is `value * value` (the first components is `value`
+    because there is no annotation on `'a` suggesting otherwise), with the
+    non-writable `*`.
+
+    Putting this all together, we define the following predicate to test for
+    writeability:
+
+    ```
+    writable : concrete_layout -> bool
+    writable(l1 * l2) = false
+    writable(l1 + l2) = false
+    writable(l1 *w l2) = writable(l1) && writable(l2)
+    writable(_) = true
+    ```
+
+    The update of a mutable field is thus allowed if *either*:
+
+    1. The layout of the field fits in a 64-bit word, *or*
+    2. The layout of the field satisfies the `writable` predicate above.
+
+    The first possibility -- for word-sized layouts -- is because a type whose
+    values fit in a 64-bit word have no threat of getting torn, and thus
+    can be updated safely. We choose *64-bit* words (not 32-bit ones) because
+    32-bit OCaml does not support multicore, and so no possibility of tearing
+    exists.
+
+    The restriction described here is called the *tearing restriction*.
+
+    The tearing restriction allows updates of our `mutable center` field,
+    for two reasons: the type `#point` is concrete and thus has layout
+    `bits32 *w bits32`, with `*w`, and a `#point` fits in a 64-bit word.
+
+    A user who wishes to update a non-atomic mutable field (with their own plan
+    for synchronization between threads) may do so with the `[@non_atomic]`
+    attribute, like so:
+
+    ```ocaml
+    type point = { x : #int32; y : #int32 }
+    module Circle : sig
+      type t : bits32 * (bits32 * bits32)
+    end = struct
+      type t = #{ radius : #int32; mutable center : #point }
+    end
+
+    type t = { mutable circle : #Circle.t; color : Color.t }
+    let f (t : t) = t.circle <- #{ ... } [@non_atomic]
+    ```
+
+    The design in this section upholds the following principle:
+
+    * Even in the presence of data races, a programmer can never observe an
+      intermediate state of a single assignment (unless they write `[@non_atomic]`).
```

Comment:
I'm afraid I don't understand your opening paragraph: when you say "How is this different here?", what is "this"? It seems you're suggesting that some aspect of the design is off (please do criticize and brainstorm!), but I'm not exactly sure what it is.

In the meantime, I will react to the counter-proposal:
* If writability is the default, is it still possible to abstract over an unboxed type with an existential variable? It would seem not -- there would have to be some way to say that a certain product is not writable.
* This effectively limits the availability of unboxed types, either preventing the abstract export of some types or placing performance-sapping burdens on the implementation of a type.
* I agree that boxing is one workaround. Yet within some teams within Jane Street, it's an unacceptable one: they need to write code that works without latency-inducing garbage collection. Indeed, the internal feedback I've gotten on this proposal already had two people essentially requesting `[@non_atomic]` (which perhaps they missed).

Reply:
Subsumed by the discussion in the main thread, feel free to mark as resolved.